### PR TITLE
time filter preprocessor

### PR DESCRIPTION
**Pull request by valeriupredoi**
simple, we may want to extend it in the future :beer:

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [ ] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [ ] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [ ] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

* * *

Closes https://github.com/ESMValGroup/ESMValCore/issues/219

### CODE REVIEW

**Comment by zklaus** [comment]
Very useful! Warrants some documentation in `doc/esmvalcore/preprocessor.rst`.

File: esmvalcore/preprocessor/_time.py [hunk 1]
```diff
@@ -553,3 +554,91 @@ def regrid_time(cube, frequency):
...
+    Raises
+    ------
+    iris.exceptions.CoordinateNotFoundError:
+        Cube does not have time coordinate.
+    """
+    try:
+        cube.coord('time')
+    except iris.exceptions.CoordinateNotFoundError:
+        err = "Cube {} does not have time coordinate".format(cube)
+        raise err
```
Comment (mathause) [inline esmvalcore/preprocessor/_time.py]: Would that not be a KeyError(err) here?

**Comment by valeriupredoi** [comment]
cheers @zklaus on my way to do it, do you reckon this func can be made even more useful in any other way? Maybe make the aggregation statistic optional for the user to choose? :beer:

File: esmvalcore/preprocessor/_time.py (hunk 1 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_time.py]: no, it is triggered by iris not finding the coordinate time, it's a specific iris exception

**Comment by valeriupredoi** [comment]
ok @mathause cheers for the suggestion, I'll see what I can plug in :beer:

**Comment by valeriupredoi** [comment]
OK @mathause made it a bit more flexible, have a looksee pls :beer:

File: esmvalcore/preprocessor/_time.py [hunk 2]
```diff
@@ -553,3 +554,109 @@ def regrid_time(cube, frequency):
...
+    -------
+    iris.cube.Cube
+        cube time-filtered using 'rolling_window'.
+
+    Raises
+    ------
+    iris.exceptions.CoordinateNotFoundError:
+        Cube does not have time coordinate.
+    NotImplementedError:
+        If filter_type is not 'lowpass'.
```
Comment (mathause) [inline esmvalcore/preprocessor/_time.py]:
```suggestion
        If filter_type is not implemented.
```

File: esmvalcore/preprocessor/_time.py [hunk 3]
```diff
@@ -553,3 +554,109 @@ def regrid_time(cube, frequency):
...
+
+    # Construct weights depending on frequency
+    # TODO implement more filters!
+    supported_filters = ['lowpass', ]
+    if filter_type in supported_filters:
+        if filter_type == 'lowpass':
+            wgts = low_pass_weights(window, 1. / span)
+    else:
+        raise NotImplementedError(
+            "Filter type {} not implemented".format(filter_type))
```
Comment (mathause) [inline esmvalcore/preprocessor/_time.py]:
```suggestion
            "Filter type {} not implemented, please choose one of {}".format(filter_type, ", ".join(supported_filters)))
```

File: esmvalcore/preprocessor/_time.py [hunk 4]
```diff
@@ -553,3 +554,109 @@ def regrid_time(cube, frequency):
...
+    try:
+        cube.coord('time')
+    except iris.exceptions.CoordinateNotFoundError:
+        logger.error("Cube %s does not have time coordinate", cube)
+        raise
+
+    # Construct weights depending on frequency
+    # TODO implement more filters!
+    supported_filters = ['lowpass', ]
+    if filter_type in supported_filters:
```
Comment (mathause) [inline esmvalcore/preprocessor/_time.py]: You don't really need to have two levels of `if`.

File: esmvalcore/preprocessor/_time.py [hunk 5]
```diff
@@ -553,3 +554,109 @@ def regrid_time(cube, frequency):
...
+    cube.html#iris.cube.Cube.rolling_window>`_
+
+    Parameters
+    ----------
+    cube: iris.cube.Cube
+        input cube.
+    window: int
+        The length of the filter window (in units of cube time coordinate).
+    span: int
+        Number of months/days (depending on data frequency) on which
```
Comment (mathause) [inline esmvalcore/preprocessor/_time.py]: I'm not sure I get the difference between `window` and `span`. Also, span could be optional for some types of filters?

File: tests/unit/preprocessor/_time/test_time.py [hunk 6]
```diff
@@ -746,6 +746,43 @@ def test_regrid_time_hour(self):
...
+            [2.44824568, 3.0603071, 3.67236852, 4.28442994, 4.89649137,
+             5.50855279, 6.12061421, 6.73267563, 7.34473705, 7.95679847,
+             8.56885989, 9.18092131, 9.79298273, 10.40504415, 11.01710557,
+             11.62916699, 12.24122841, 12.85328983]
+        )
+        assert_array_almost_equal(filtered_cube.data, expected_data)
+        assert len(filtered_cube.coord('time').points) == 18
+
+    def test_timeseries_filter_timecoord(self):
+        """Test a not implemnted filter."""
```
Comment (mathause) [inline tests/unit/preprocessor/_time/test_time.py]:
```suggestion
        """Test missing time axis."""
```

File: tests/unit/preprocessor/_time/test_time.py [hunk 7]
```diff
@@ -746,6 +746,43 @@ def test_regrid_time_hour(self):
...
+        filtered_cube = timeseries_filter(self.cube, 7, 14)
+        expected_data = np.array(
+            [2.44824568, 3.0603071, 3.67236852, 4.28442994, 4.89649137,
+             5.50855279, 6.12061421, 6.73267563, 7.34473705, 7.95679847,
+             8.56885989, 9.18092131, 9.79298273, 10.40504415, 11.01710557,
+             11.62916699, 12.24122841, 12.85328983]
+        )
+        assert_array_almost_equal(filtered_cube.data, expected_data)
+        assert len(filtered_cube.coord('time').points) == 18
+
```
Comment (mathause) [inline tests/unit/preprocessor/_time/test_time.py]: maybe add test with explicit selection of the filter:

``` python

filtered_cube = timeseries_filter(self.cube, 7, 14, filter_type="lowpass")
assert_array_almost_equal(filtered_cube.data, expected_data)
assert len(filtered_cube.coord('time').points) == 18

```

File: esmvalcore/preprocessor/_time.py (hunk 4 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_time.py]: that assumes there are different functions for different filter types

File: tests/unit/preprocessor/_time/test_time.py (hunk 7 above)
Comment (valeriupredoi) [reply tests/unit/preprocessor/_time/test_time.py]: good call, we should test the other (future) filter types too; done, cheers :beer:

**Comment by bouweandela** [comment]
@valeriupredoi Can you please use the pull request template? It makes it much easier for everyone to see where in the review process a pull request is and what still needs to be done.

**Comment by bouweandela** [comment]
@valeriupredoi What's the status of this? Is it ready for review/testing? Can you fix the merge conflict and Codacy issues?

**Comment by valeriupredoi** [comment]
Yup, this is ready to go, will fix the conflicts on my return t work next week :beer:

**Comment by valeriupredoi** [comment]
Le @bouweandela fixed les conflicts :beer:

File: doc/esmvalcore/preprocessor.rst [hunk 8]
```diff
@@ -905,6 +906,21 @@ set manually by the user in recipe.
...
+---------------------
+
+This function allows the user to apply a filter to the timeseries data. This filter may be
+of the user's choice (currently only the ``low-pass`` Lanczos filter is implemented); the
+implementation is inspired by this `iris example
+<https://scitools.org.uk/iris/docs/latest/examples/General/
+SOI_filtering.html?highlight=running%20mean>`_ and uses aggregation via a
+`rolling window <https://scitools.org.uk/iris/docs/v2.0/iris/iris/cube.html#iris.cube.Cube.rolling_window>`_.
+
+See also :func:`esmvalcore.preprocessor.low_pass_weights`.
```
Comment (bouweandela) [inline doc/esmvalcore/preprocessor.rst]:
```suggestion
```
This doesn't look like a preprocessor function

File: doc/esmvalcore/preprocessor.rst [hunk 9]
```diff
@@ -905,6 +906,21 @@ set manually by the user in recipe.
...
+``timeseries_filter``
+---------------------
+
+This function allows the user to apply a filter to the timeseries data. This filter may be
+of the user's choice (currently only the ``low-pass`` Lanczos filter is implemented); the
+implementation is inspired by this `iris example
+<https://scitools.org.uk/iris/docs/latest/examples/General/
+SOI_filtering.html?highlight=running%20mean>`_ and uses aggregation via a
+`rolling window <https://scitools.org.uk/iris/docs/v2.0/iris/iris/cube.html#iris.cube.Cube.rolling_window>`_.
+
```
Comment (bouweandela) [inline doc/esmvalcore/preprocessor.rst]: Can you make this a bit nicer? I am missing the Parameters and Examples sections, see e.g. https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html#climate-statistics

**Comment by bouweandela** [comment]
> This pull request has a descriptive title that can be used in a changelog

Can you please try to improve this?

**Comment by bouweandela** [comment]
@valeriupredoi Any news on this?

**Comment by valeriupredoi** [comment]
Whoops, cheers for the nudge Bouwe, will fix in the morrow :beer:

**Comment by valeriupredoi** [comment]
all yours @bouweandela :beer:

**Comment by valeriupredoi** [comment]
cheers B-man! Mattia can go ahead and test at will, I'll see what the Codacy was nagging about, she's running at the moment :beer:

**Comment by valeriupredoi** [comment]
aw she just finished - nope, forget about it, not gonna bother with `single character variable is not something I like` :grin:

**Comment by mattiarighi** [comment]
Single character variables are a nightmare when you search for them in the code.
So I strongly support fixing those issues :muscle: 

Sorry mate.

**Comment by mattiarighi** [comment]
You can't use Capitalized words for variables... :rofl:

**Pull request merged**